### ISSUE TITLE
RFC: "Unqualified" imports/lets

### ISSUE THREAD

**Issue by SiriusStarr** [dhall-lang/dhall-lang#863]
**Rationale**

As we all know, most all Dhall frameworks (for example, the Prelude) consist of records (`package.dhall` in the prelude) of functions, values, types, etc., with the equivalent of "importing" a module doing something like
```
let Prelude = https://prelude.dhall-lang.org/package.dhall
in Prelude.List.null
```

This for the most part works out very nicely but has one downside, namely the need to prefix everything with the binding in question.

I do a lot of text templating with Dhall, which lends itself to having a lot of short helper functions/etc that are used in interpolation, and this need to prefix everything gets very annoying to read very fast.  I often end up doing something like
```
let e = ./Environment/package.dhall
in ''
   This is text with interpolation.${e.br}And some ${e.emph "formatting"}.
   ''
```
Is it the absolute end of the world?  No.   But it is cumbersome and harms legibility of the text in question.  And it would be really nice to be able to do `This is text with interpolation.${br}And some ${emph "formatting"}.` without having to write out a thousand `let`s at the top of the file manually.

**Proposal**

This proposal is simple: a new syntax construct for bringing into the namespace the fields in a record, rather than the record itself.  Some possible ideas (but obviously open to others):
```
expose https://prelude.dhall-lang.org/package.dhall
in List.null
```
```
import https://prelude.dhall-lang.org/package.dhall
in List.null
```

This could be combined with record projection (already in the language) to be able to do things like:
```
expose (https://prelude.dhall-lang.org/package.dhall).{List, Text}
in {null = List.null, concat = Text.concat}
```

I've talked only exclusively about imports here, since that's where it'd be useful, but there is obviously no reason this would need to be restricted, e.g. it would be valid to do (though why you would...):
```
let Record = { a = 1, b = 2}
expose Record
in a
```

**Implementation**

In all cases, this could simply desugar to multiple `let` bindings; no need for any other changes to the language.

Trying to expose an expression that is not a record would be a type error.

As it's only syntactic sugar for `let`s, shadowing would work in the same way, e.g.
```
expose https://prelude.dhall-lang.org/package.dhall

let List = 3

in List
```
would evaluate to `3 : Natural`.

**Backwards Compatibility**

This would likely be a breaking change only inasmuch as the new keyword would be a reserved identifier.

**Comment by sjakobi** [comment]
> In all cases, this could simply desugar to multiple `let` bindings; no need for any other changes to the language.

Can you explain the desugaring in some more detail, with examples?

Can you compare your proposal to https://github.com/dhall-lang/dhall-lang/issues/74?

**Comment by SiriusStarr** [comment]
> Can you explain the desugaring in some more detail, with examples?

This obviously would have to occur after resolving the import (if it's one), since before, we aren't sure that what we're exposing is a record.  The basic idea is that:
```
expose { a = x, b = y, c = z}
```
becomes
```
let a = x
let b = y
let c = z
```

The only issue I foresee is `expose {=}`, since `expose {=} in a` can't become just `in a` (since that's a syntax error). 

> Can you compare your proposal to #74?

#74 (to my reading) has no option to expose *everything* from a record and thus requires manually typing out every single field you want to expose, at which point you might as well write all the `let`s by hand.  Unless I'm misreading something?  This would render #74 unnecessary, since you could just use record projection to get the same behavior, e.g. instead of 
```
let { map, filter, span } = http://prelude.dhall/List
```
do
```
expose (http://prelude.dhall/List).{ map, filter, span }
```

**Comment by ari-becker** [comment]
@SiriusStarr there's a problem with a simple desugar.

```dhall
expose { a = 1, b = 2, c = 3 }
expose { d = a , e = b , f = c }
in { d , e , f }
```
would then denormalize to `{ d = 1, e = 2, f = 3 }` when really it should result in a type error since `{ d = a, e = b, f = c }` is not self-contained.

**Comment by SiriusStarr** [comment]
> @SiriusStarr there's a problem with a simple desugar.
> 
> ```haskell
> expose { a = 1, b = 2, c = 3 }
> expose { d = a , e = b , f = c }
> in { d , e , f }
> ```
> 
> would then denormalize to `{ d = 1, e = 2, f = 3 }` when really it should result in a type error since `{ d = a, e = b, f = c }` is not self-contained.

It's not immediately clear to me that that *should* be a type error.  You can use `let` bindings in subsequent `let` bindings, so I don't understand why `expose` should be self-contained when `let`s aren't?

**Comment by Gabriel439** [comment]
@SiriusStarr: Normally, I would be mildly against supporting "open" import, but when I saw that you could combine it with record projection to create explicit imports:

```dhall
let Prelude = https://prelude.dhall-lang.org/List/package.dhall

expose Prelude.List.{ map, filter, span }

…
```

... then I changed my mind because that is pretty elegant, so I approve of making this change.

I think the main thing I'd like to bikeshed a little bit about is the name for the keyword.  Several languages have an existing operator like this and I'd like to reuse prior art when naming if possible.  For example, Nix calls this `with` and Agda calls this `open`.

**Comment by ari-becker** [comment]
@SiriusStarr if you instead split out the two `expose` lines into their own files, i.e. `expose ./abc.dhall` and `expose ./def.dhall`, clearly `def.dhall` would not parse on its own.

`let` statements are different because they are implicitly joined with a following `in` - it is the combination of the `let` with the `in` which is self-contained.

I'm not opposed to the idea of an unqualified import - I just think that the import should be self-contained, like any other import.

**Comment by SiriusStarr** [comment]
> I think the main thing I'd like to bikeshed a little bit about is the name for the keyword. Several languages have an existing operator like this and I'd like to reuse prior art when naming if possible. For example, Nix calls this `with` and Agda calls this `open`.

Yeah, definitely.  I'm not set on `expose`; I just needed a placeholder keyword to use in discussion.

@ari-becker Wouldn't the import fail here just as well as it would if you did `let x = ./def.dhall`?  The import needs to be resolved before desugaring can happen, and presumably that import will throw a type error.

> `let` statements are different because they are implicitly joined with a following `in`

In my examples I proposed a similar syntax for this case, with it being `expose x in y`

**Comment by Gabriel439** [comment]
@ari-becker: I don't believe that @SiriusStarr is proposing that an imported record can refer to variables in scope at the point of the `expose` keyword.  My understanding of this proposal is that if you had a file whose contents were:

```haskell
{ x = y }
```

That would always fail to be imported, regardless of whether or not it is used within an `expose` keyword.  In fact, `expose` wouldn't really care how it got its argument record because `expose` is evaluated at normalization time and not import resolution time.

**Comment by sjakobi** [comment]
I think this has the potential to create some confusing error messages.

Say you start with

```dhall
expose Prelude.List
in map …
```

Then you add another `expose`

```dhall
expose Prelude.List
expose Prelude.Optional
in map …
```

and suddenly you get an error about some `List` that should be an `Optional`.

https://github.com/dhall-lang/dhall-lang/issues/74 makes errors like this one easier to debug since all bindings are explicit. So maybe we could make the projection obligatory? Then we'd have

```dhall
expose Prelude.List.{ map }
expose Prelude.Optional.{ default }
in map …
```

and shadowed bindings would be easier to detect.

**Comment by sjakobi** [comment]
I'd also expect that this would be used for destructuring records, e.g.

```
let printBigRecord = \(b : BigRecord) -> expose b.{ x, y } in …
```

But then `BigRecord` gets another field `z`, `printBigRecord` doesn't get updated, and we have a bug.

So I think it would be good to have some kind of exhaustiveness check, as discussed in #74.

**Comment by SiriusStarr** [comment]
> So maybe we could make the projection obligatory?

That defeats the purpose of the proposal, not having to individually type out the names of hundred helper functions I'm trying to import.  You can always use projection to limit things if you don't need them, so I don't see how the errors would be confusing unless you intentionally go around importing tons of things unqualified.  (On a related note, perhaps Dhall should warn about overwritten `let` bindings?  That would presumably render this non-confusing.)

I mean, one possible compromise would be to require projection but allow a "wildcard" projection, e.g.
```
expose Prelude.List.{*}
```
or something (which requires one to *explicitly* expose everything and serves as a sort of "are you sure you meant to do that?" check); I avoided proposing this originally since it would require additional changes to the language.

> I'd also expect that this would be used for destructuring records, e.g.
> 
> ```
> let printBigRecord = \(b : BigRecord) -> expose b.{ x, y } in …
> ```
> 
> But then `BigRecord` gets another field `z`, `printBigRecord` doesn't get updated, and we have a bug.
> 
> So I think it would be good to have some kind of exhaustiveness check, as discussed in #74.

At that point, it stops being record projection and would have to use different syntax, presumably, which requires a more complex proposal (since record projection doesn't have an exhaustiveness check (and probably shouldn't)).  It's not clear to me that this is in any way introduced or compounded by this proposal.  How is this any different form the current case of:
```
let printBigRecord = \(b : BigRecord) -> show b.x ++ show b.y
```
or
```
let printBigRecord = \(b : BigRecord) -> let x = b.x let y = b.y in show x ++ show y
```
(or whatever the print is)?  Presumably the bug would still occur in those cases, since adding a field to `BigRecord` isn't going to result in any error/warning.

**Comment by f-f** [comment]
> #74 makes errors like this one easier to debug since all bindings are explicit. So maybe we could make the projection obligatory?

Yeah I would also say that we should require the projection to be there. 

My single biggest frustration about Haskell is the fact that one can have open imports, which makes it terrible when it comes to reading new code and understanding which symbols come from where. So I'd be strongly against adding open imports here.

**Comment by sjakobi** [comment]
@SiriusStarr Indeed the current way of destructuring a record by record selection doesn't offer exhaustiveness checking. But I think that's something that we'll eventually want. So if we'd now offer a destructuring mechanism that doesn't include an exhaustiveness check, I think it might be harder to add one later, since people might then take the "dangerous" shortcut for convenience.

**Comment by SiriusStarr** [comment]
> the current way of destructuring a record by record selection doesn't offer exhaustiveness checking. But I think that's something that we'll eventually want.

:woman_shrugging: I mean, if people want to go to the work of adding completely novel syntax for destructuring records, I'm fine with that **as long as there's a wildcard option** (it's fine if it has to be explicitly done and not the default).  I was just trying to be parsimonious and utilize existing constructs in the language so as not to have a bunch of overhead.

**Comment by Gabriel439** [comment]
@sjakobi: I no longer think the absence of exhaustivity checking is a requirement for destructuring records.  Specifically, what convinced me is the implementation of closed imports in terms of open imports + record projection.  If exhaustivity checking is not required for record projection, then it seems to me like it shouldn't be required for destructuring a record either.

@sjakobi @f-f: Perhaps a compromise is that we can have `dhall lint` reject programs with either (A) any open imports or (B) more than one open import (which is the same rule PureScript uses).

**Comment by f-f** [comment]
Yeah thinking more about it, allowing one and only one open import sounds sensible. Though I think the check (and eventual rejection of the program) should be performed at the import resolution phase, as not everyone uses `dhall lint`

**Comment by SiriusStarr** [comment]
> Though I think the check (and eventual rejection of the program) should be performed at the import resolution phase

Hmm, I guess the question there is should multiple open imports be forbidden by the standard or just strongly discouraged by the tooling/community?  I shy away from the former simply because it adds a largely unnecessary degree of complexity to Dhall implementations.  (And who knows, maybe there is some legitimate usecase for multiple open imports that we're not thinking of?)

**Comment by f-f** [comment]
@SiriusStarr: the problem with optional/not-enforced things is that they end up being ignored. 

A very good example of this in my experience is "optional typing": if it's not enforced by the compiler it will be very hard to actually type a whole system.
Python has optional typing and I could use that to write my programs, but I choose Haskell/Dhall and other languages with strong typing because the compiler enforces the use of a good typesystem in order to prevent unnecessary pain to yourself and others.

On the same line, in my experience as an engineer I got traumatized by open imports: every time I try to understand a giant module with tons of open imports it always feels horrible because I have no idea where stuff comes from and there's no shortcut to take to find out (except for "go to definition", but that's still painfully slow)

To summarize:
- not allowing open imports is a good UX maintainability-wise because you always know where things come from
- allowing one open import (most likely the Prelude) is still sensible
- allowing multiple open imports has the potential to traumatize other engineers that will have to maintain our code in the future, and I don't really want to take this risk

Given the fact that Dhall advertises itself as "maintainable configuration language", I think that to live up to that aspiration we should really enforce this with no possibility for opting out, just like our type system is.

**Comment by SiriusStarr** [comment]
Hmm, I guess I see it more as Elm vs Haskell.  `unsafeCoerce` exists but doesn't invalidate the type-safety of Haskell, and I prefer the language for it trusting me with the power to (for example) write partial functions.  But the compiler still yells at me repeatedly if I do, so I can't do it by accident.

Regardless, I'm certainly not going to pick this hill to die on if people feel strongly about it.  I was just trying to make life easier for those writing implementations of Dhall.  If there's a general consensus that only allowing one open import is good and it's not horrible to implement then that's fine.

**Comment by sjakobi** [comment]
This discussion reminds me a bit of another classic post by Gabriel: ["Worst practices should be hard"](http://www.haskellforall.com/2016/04/worst-practices-should-be-hard.html)

**Comment by joneshf** [comment]
> My single biggest frustration about Haskell is the fact that one can have open imports, which makes it terrible when it comes to reading new code and understanding which symbols come from where. So I'd be strongly against adding open imports here.

Ditto.

This proposal is a big -1 for me. Dumping things into the global scope makes everything much more complex (as this issue is already showing). Even if you buy into making implementations more complex, the effect for end-users is the real huge negative. Someone that is intimately familiar with Dhall can navigate this stuff easily. Someone who is new, or doesn't spend all their time learning the intricacies of the language, has orders of magnitude more difficulty with this stuff.

Hiding syntactic complexity is almost never worth the price you pay for it.

**Comment by SiriusStarr** [comment]
> Hiding syntactic complexity is almost never worth the price you pay for it.

I mean, by that logic shouldn't everything be written in assembly?  Certainly the entirety of FP is a huge degree of abstraction/hiding of syntactic complexity, since computers are imperative at the hardware level.  It's not a matter of syntactic complexity, because there is nothing syntactically complex about:
```
let a = ...
let b = ...
let c = ...
let d = ...
let e = ...
```
It's just boilerplate that A: costs me time to have to write and B: is then cruft I have to stare at when I'm trying to actually find what's actually relevant to code.

**Comment by f-f** [comment]
Meta-note, quoting from https://github.com/dhall-lang/dhall-lang/issues/260#issuecomment-433992627:

> In the interest of avoiding hostility I'm going to ask that people refrain from using GitHub reactions (both positive and negative) to weigh in on a divided issue like this one since they can easily be misconstrued and turn this into a popularity contest instead of a technical discussion.

**Comment by f-f** [comment]
@SiriusStarr I agree that the sequence of `let`s there doesn't look too good, and in fact this would be solved by #74 where you could write
```dhall
let { a, b, c, d, e } = ..
```

**Comment by SiriusStarr** [comment]
@f-f 

I honestly don't have the spoons for this at the moment, so forgive me if this is terse.

If #74 gets standardized, great, but it's been 2 years and it hasn't been, so I was just trying to do something elegant using existing constructs in the language (record projection).  #74 would be better than nothing, certainly, if less than ideal.  If it's generally agreed this is a horrible idea, I'll close it and move on.

**Comment by Gabriel439** [comment]
@f-f @sjakobi @joneshf: What if we enforced the "at most one open import" rule at the language level?

**Comment by sjakobi** [comment]
> What if we enforced the "at most one open import" rule at the language level?

I think that would be an improvement on the original proposal, but I still feel that exhaustiveness checking is important.

In my opinion it would be better to standardize #74 (with exhaustiveness checking) first, and if that turns out not to be sufficiently convenient, we can consider adding wildcards, escape hatches for the exhaustiveness checking, etc.

**Comment by Nadrieril** [comment]
> What if we enforced the "at most one open import" rule at the language level?

That would prevent a lot of legitimate uses for a nested open import local to just one expression. And more generally it would mean that "manually inlining" an import could make code stop building, which I don't like.
I'm more in favor of #74 too.

**Comment by f-f** [comment]
I'm on board with what @sjakobi and @Nadrieril suggested, which is to try out #74 first and then see where to go from there

**Comment by Gabriel439** [comment]
@Nadrieril @f-f @sjakobi: Alright, I'll support #74 on one condition: that one of its proponents volunteers to standardize it.  The reason why is that I believe that in a healthy language evolution process we should reward those willing to put in the work (in this case standardization-related work).

@SiriusStarr seems willing to put in the work standardize this, so before rejecting his proposal one of us should reciprocate by volunteering to work on the preferred alternative (#74 in this case).

**Comment by sjakobi** [comment]
> @Nadrieril @f-f @sjakobi: Alright, I'll support #74 on one condition: that one of its proponents volunteers to standardize it. The reason why is that I believe that in a healthy language evolution process we should reward those willing to put in the work (in this case standardization-related work).

I think that's a very fair and prudent way of putting things. :+1: 

Right now #74 needs more discussion and I fear that it could become rather complex. I'm willing to at least _help_ standardize it, although I'm not sure that I have the chops to do it alone.

**Comment by Profpatsch** [comment]
`with` in nix breaks lots of things (and has confusing shadowing semantics), I don’t know how it would be in dhall. But in general imports that don’t explicitely bring symbols into scope hurt readability of a language very much. Also greppability.

**Comment by Twisol** [comment]
For a domain I'm interested in, it is hard to support convenient authoring of domain entities without either adding lots of imports or indexing into a record. The former adds a large amount of clutter for this domain, and the latter adds noise to the specification of the entity itself. #74 doesn't really help with this use-case, because each of the many domain concepts in use must still be imported individually.

The alternative is essentially creating a "header" file with a bunch of `let` statements, terminated by an `in`, and concatenating this file with the entity specification. This has the same effect as defining a custom prelude as in this ticket, so it certainly _works_, but it isn't very elegant.

Having a construct like `expose ./domain.dhall in ...` would be a significant improvement, because the file itself declares where the ambient definitions are coming from, so you can at least have a hope of tracing down any definitions. Earlier discussion in this thread mentioned limiting to one use of `expose` in a file, and I don't see any problem with that at all.

I agree with other commentors that implicit imports generally impair readability. However, I think there is a legitimate use-case where implicit imports make more sense, particularly regarding specifications in rich domains with lots of concepts already given at the ground floor.

**Comment by Gabriel439** [comment]
Was there anybody that opposed the idea of "at most one open import"?  It seems like that would address most people's use case for open imports, including @Twisol's use case

**Comment by sjakobi** [comment]
> Was there anybody that opposed the idea of "at most one open import"?

Well I was opposed to the whole thing really, mostly because I think that users will end up using open imports when they ought to use some construct that comes with an exhaustiveness check, like what I've been arguing for in #74. 

I've become somewhat less convinced though of my line of thinking that language constructs that _can_ be misused shouldn't be added at all.

Also, since there hasn't been much progress on standardizing #74, I think I shouldn't try to block this proposal here further.

One question though: Would it be reasonable to limit open imports to "concrete values", as opposed to values where we only know the type? That way they would be harder to misuse in contexts where something like #74 would be better.

What I mean is that

```dhall
expose ./foo in ...
```

or

```dhall
expose { a = 1, b = 2 } in ...
```

should be legal, but e.g.

```dhall
\(x : Foo) -> expose x in ...
```

should not.

**Comment by sjakobi** [comment]
> Would it be reasonable to limit open imports to "concrete values", as opposed to values where we only know the type?

I _think_ what I'm asking for here is that the argument to `expose` should type-check with an empty context.

**Comment by Nadrieril** [comment]
So you want to allow
```dhall
\(x : Foo) -> expose { a = 1 } in ...
```
but not
```dhall
\(x : Foo) -> expose { a = x } in ...
```
?

**Comment by sjakobi** [comment]
@Nadrieril That would be a consequence of typechecking the record with an empty context.

**Comment by Gabriel439** [comment]
It's not clear to me how restricting the context solves the main issue with open imports: not knowing where an identifier originates from

**Comment by sjakobi** [comment]
> It's not clear to me how restricting the context solves the main issue with open imports: not knowing where an identifier originates from

It doesn't really. My idea is to restrict `expose` to use-cases where the shape and contents of the record are likely to be fairly constant. In such a use-case the "identifier of unknown origin problem" seems less grave – users just need to memorize the exposed "domain" once.

OTOH, when we have a record of unknown contents, it seems more likely that we'll also want exhaustiveness checking, as provided by the proposal in #74. 

Does this make any sense?! ;)

**Comment by Nadrieril** [comment]
We can combine the two ideas: allow open imports at most once in a file, and not under lambdas.

**Comment by Gabriel439** [comment]
@sjakobi: I don't believe the empty context restriction would change the user's ability to reason about which fields were brought into the scope.  The reason why is that the fields of the record would already need to be known statically in order to desugar `expose` to equivalent `let` bindings.  As far as I can tell, the new difference the restriction introduces is that now the values of the fields cannot refer to things in context, but I don't understand why we would want to superimpose that additional restriction.

**Comment by lisael** [comment]
I'd like to add a further restriction on open imports: that it must be in first position of the let bindings (i.e. at the root of the file's context) to avoid accidental shadowing of previously bound names. This improve legibility by a lot even for someone new to dhall or to the particular code, solves the original problem without unleashing uncontrolled uses, and mitigates @joneshf 's [objections]( https://github.com/dhall-lang/dhall-lang/issues/863#issuecomment-567671838) (which I tend to agree on, but I also agree with the original issue).

Also, I like the idea of being explicit about the fact it's an open import (@SiriusStarr proposed `.{*}` for this), but it's less important, to me.

legit:
```dhall
expose ./foo.dhall.{*}
let bar = 1
in  qux  -- I'm certain that `qux` is defined in ./foo.dhall
```

also legit:
```dhall
expose ./foo.dhall.{*}
let bar = 1
expose ./baz.dhall.{bar}
in  bar + qux  -- `qux` has to come from foo.dhall and `bar` was explicitely shadowed 
```

not legit:
```dhall
expose ./baz.dhall.{bar, qux}
expose ./foo.dhall.{*}
in bar  -- ?? no way to tell at first glance where `bar` is defined
```

**Comment by SiriusStarr** [comment]
I think that's a good idea, since it guarantees there's no potential to override `let`s in the file accidentally (but still allows you to override something from the open import if necessary).  It also organically effects the "only one open import per file" restriction without having to actually make that an explicit restriction (if it must be the first in the file, there can be only one).

**Comment by Gabriel439** [comment]
@lisael: If we introduce new `{*}` syntax for this, then do we still need the `expose` keyword?  I was thinking we could instead use a syntax similar to the proposed idea for unpacking records (https://github.com/dhall-lang/dhall-lang/issues/74):

```dhall
let {*} = ./foo.dhall

…
```

**Comment by Nadrieril** [comment]
Note that we would still be able to emulate multiple open imports with
```dhall
let {*} = ./foo.dhall // ./bar.dhall

…
```
I think that's positive actually, because we might want for example the contents of both `Prelude.JSON` and `Prelude.Map` in scope, and the syntax would discourage abusing it.

**Comment by Twisol** [comment]
> ```dhall
> let {*} = ./foo.dhall // ./bar.dhall
> ```

That's a great point, and I actually really like how this looks _and_ how it works. You get the benefits of a "single open import" in that there's only one place to look, but you also have the flexibility of multiple open imports in that you can still compose your environment easily out of multiple modules. Plus, the syntax is a special case of destructuring bindings, rather than having completely separate syntax for something that can only be used _exactly once_.

**Comment by lisael** [comment]
@Gabriel439 the `expose` keyword bothers me too. I don't like profusion of keywords in languages.

I was thinking about something like that:

```hs 
let _ = ./foo.dhall as Expose
```
It looks weird. The `as` keyword is an import modifier and belongs to the resolve pass. Here we need to express the fact that we are modifying the let binding at the type checking pass. I dropped this idea for now but maybe it can fuel someone's creativity...

Regarding `let {*} = ... `, It looks nice and feels at the right place. I'm just concerned that `{*}` could be useful for something else. If we use it, for some sort of unpacking for instance, the two radically different significations may be confusing for the user.

Anything disallowed in binding names can be used here.

What about
```hs
let . = ./foo.dhall
```
pros:
- it almost universally denotes "here" (unix paths, python imports...)

cons:
- it's just a tiny dot. less legible.

**Comment by lisael** [comment]
@Nadrieril, @Twisol : 

> ```haskell
> let {*} = ./foo.dhall // ./bar.dhall
> ```

I was scared at first because it kind of defeats the purpose of the "Only one open import per file" rule. Anyway, even if we don't allow this, the user can create a chain of ad-hoc files to emulate this behaviour, which is worse.

>It's also more inspectable, since if you're wondering what bindings you're getting, you can just evaluate the right-hand side of the let on its own. 

This changed my mind. I find this pretty elegant, now.

**Comment by winitzki** [comment]
The conclusion seems to be that we don't need an `expose` keyword; instead we want to implement https://github.com/dhall-lang/dhall-lang/issues/74 with the syntax `let {*} = ...` and with the restriction that a Dhall file may have at most one `let {*} = ...` at top level and it must be the first `let` expression. There may not be any more `let {*}` anywhere in the file.
